clang/lib/CIR/Dialect/Transforms/LoweringPrepare.cpp

```diff
@@ -1120,6 +1125,78 @@ void LoweringPreparePass::lowerIterEndOp(IterEndOp op) {
   op.erase();
 }
 
+void LoweringPreparePass::lowerStdInitializerListOp(StdInitializerListOp op) {
+  auto loc = op.getLoc();
+  cir::CIRDataLayout dataLayout(theModule);
+  auto ptrDiffTy = cir::IntType::get(&getContext(), 64, /*isSigned=*/false);
+
+  auto args = op.getArgs();
+
+  auto stdInitializerListType =
+      mlir::cast<cir::StructType>(op.getInitList().getType().getPointee());
+  clang::RecordDecl::field_range stdInitializerListFields =
+      stdInitializerListType.getAst().getRawDecl()->fields();
+  assert(llvm::range_size(stdInitializerListFields) == 2U);
+
+  mlir::Type elementType =
+      mlir::cast<cir::PointerType>(stdInitializerListType.getMembers()[0])
+          .getPointee();
+  auto tempArrayType =
+      cir::ArrayType::get(&getContext(), elementType, args.size());
+  auto alignment = mlir::IntegerAttr::get(
+      mlir::IntegerType::get(&getContext(), 64),
+      dataLayout.getPrefTypeAlign(tempArrayType).value());
+
+  CIRBaseBuilderTy builder(getContext());
+  builder.setInsertionPointAfter(op);
+
+  mlir::Value arrayPtr = builder.createAlloca(
```

Comment:
I doubt [this comment](https://github.com/llvm/clangir/pull/1121#discussion_r1842779713) is resolved. Please don't mark "resolved" by yourself in this repository, feel free to leave them for the duty of reviewers ; )

In CodeGen/CIRGen, we use a helper function `CreateTempAlloca` to handle it, and it returns `Address` abstraction.

But it seems there already exists some non-entry `alloca`s in LP (e.g. in `lowerArrayDtorCtorIntoLoop`). Maybe we can just mark FIXME here and file an issue for it?

Reply:
I am not sure about CIR. the review comment is for LLVMIR instead of CIR.
If we do alloca later in CIR, lowered alloca in LLVMIR are still at the beginning of function. (see test cases)

Reply:
Indeed! That must be processed by #887 . Thanks for the clarification.

Reply:
FWIW, you could find the defining op of the first op argument to get the init_list alloca and place your new alloca next to it, but HoistAllocas is doing the job already, so this is fine.